src/goto-instrument/contracts/contract_wrangler.cpp

```diff
@@ -0,0 +1,391 @@
+/*******************************************************************\
+
+Module: Parse and annotate contracts from configuration files
+
+Author: Qinheping Hu
+
+Date: June 2023
+
+\*******************************************************************/
+
+/// \file
+/// Parse and annotate contracts
+
+#include "contract_wrangler.h"
+
+#include <util/expr_iterator.h>
+#include <util/format_expr.h>
+#include <util/pointer_expr.h>
+#include <util/simplify_expr.h>
+#include <util/std_types.h>
+
+#include <goto-programs/remove_unused_functions.h>
+
+#include <ansi-c/ansi_c_parser.h>
+#include <ansi-c/ansi_c_typecheck.h>
+#include <goto-instrument/contracts/utils.h>
+
+contract_wranglert::contract_wranglert(
+  goto_modelt &goto_model,
+  const std::string &file_name,
+  messaget &log)
+  : ns(goto_model.symbol_table),
+    goto_model(goto_model),
+    symbol_table(goto_model.symbol_table),
+    goto_functions(goto_model.goto_functions),
+    log(log)
+{
+  jsont configuration;
+
+  if(parse_json(file_name, log.get_message_handler(), configuration))
+    throw deserialization_exceptiont(file_name + " is not a valid JSON file");
+
+  configure_functions(configuration);
+
+  // Mangle loop contract function by function.
+  // We match function with a regex. There should one and only one function
+  // with name matched by the regex.
+  for(const auto &function_entry : this->functions)
+  {
+    bool function_found = 0;
+
+    for(const auto &function : goto_model.goto_functions.function_map)
+    {
+      // We find a matched function.
+      if(std::regex_match(function.first.c_str(), function_entry.first))
+      {
+        if(function_found)
+          throw deserialization_exceptiont(
+            "function regex" + function_entry.second.regex_str +
+            " match more than one function");
+
+        function_found = 1;
+
+        // Mangle all loop contracts in the function.
+        for(const auto &loop_entry : function_entry.second.loop_contract)
+        {
+          mangle(loop_entry, function.first);
+        }
+      }
+    }
+
+    if(!function_found)
+      throw deserialization_exceptiont(
+        "function " + function_entry.second.regex_str + " match no function");
+  }
+}
+
+void contract_wranglert::add_builtin_pointer_function_symbol(
+  std::string function_name,
+  const unsigned num_of_args)
+{
+  // Already exist.
+  if(symbol_table.lookup(CPROVER_PREFIX + function_name) != nullptr)
+    return;
+
+  code_typet::parameterst parameters;
+  for(unsigned i = 0; i < num_of_args; i++)
+  {
+    parameters.emplace_back(pointer_typet(signedbv_typet(32), 32));
```

Comment:
This hard-coding of bitwidths is likely wrong. I'm pretty sure you want to use the functions from `util/c_types.h` (`signed_int_type` and `pointer_type`).

Reply:
Yes. I use `pointer_type(void_type())` instead.